airbrakes/hardware/imu.py

```diff
@@ -123,36 +203,93 @@ def _fetch_data_loop(self, port: str, frequency: int, _: bool):
         """
-        self._running.value = False
-        # Waits for the process to finish before stopping it
-        self._data_fetch_process.join()
+        # Calculate the interval between readings based on frequency
+        # TODO: reading takes a little bit of time, so we should probably subtract a few milliseconds from the interval
+        interval = 1.0 / frequency
+
+        with log_file_path.open(newline="") as csvfile:
+            reader = csv.DictReader(csvfile)
+
+            for row in reader:
+                # Check if the process should stop
+                if not self._running.value:
+                    break
+
+                # Create the data packet based on the row
+                if row.get("scaledAccelX") is not None:
+                    imu_data_packet = RawDataPacket(int(row["timestamp"]))
+                    imu_data_packet.scaledAccelX = float(row.get("scaledAccelX"))
+                    imu_data_packet.scaledAccelY = float(row.get("scaledAccelY"))
+                    imu_data_packet.scaledAccelZ = float(row.get("scaledAccelZ"))
+                    imu_data_packet.scaledGyroX = float(row.get("scaledGyroX"))
+                    imu_data_packet.scaledGyroY = float(row.get("scaledGyroY"))
+                    imu_data_packet.scaledAccelZ = float(row.get("scaledGyroZ"))
+                elif row.get("estCompensatedAccelX") is not None:
+                    imu_data_packet = EstimatedDataPacket(int(row["timestamp"]))
+                    imu_data_packet.estCompensatedAccelX = float(row.get("estCompensatedAccelX"))
+                    imu_data_packet.estCompensatedAccelY = float(row.get("estCompensatedAccelY"))
+                    imu_data_packet.estCompensatedAccelZ = float(row.get("estCompensatedAccelZ"))
+                    imu_data_packet.estAngularRateX = float(row.get("estAngularRateX"))
+                    imu_data_packet.estAngularRateY = float(row.get("estAngularRateY"))
+                    imu_data_packet.estAngularRateZ = float(row.get("estAngularRateZ"))
+                    imu_data_packet.estOrientQuaternionW = float(row.get("estOrientQuaternionW"))
+                    imu_data_packet.estOrientQuaternionX = float(row.get("estOrientQuaternionX"))
+                    imu_data_packet.estOrientQuaternionY = float(row.get("estOrientQuaternionY"))
+                    imu_data_packet.estOrientQuaternionZ = float(row.get("estOrientQuaternionZ"))
+                    imu_data_packet.estPressureAlt = float(row.get("estPressureAlt"))
+                    imu_data_packet.estAttitudeUncertQuaternionW = float(row.get("estAttitudeUncertQuaternionW"))
+                    imu_data_packet.estAttitudeUncertQuaternionX = float(row.get("estAttitudeUncertQuaternionX"))
+                    imu_data_packet.estAttitudeUncertQuaternionY = float(row.get("estAttitudeUncertQuaternionY"))
+                    imu_data_packet.estAttitudeUncertQuaternionZ = float(row.get("estAttitudeUncertQuaternionZ"))
```

Comment:
why not a for loop

Reply:
cause this was easy and each row has more data than just whats in the packet